src/python/pants/option/options_bootstrapper.py

```diff
@@ -66,68 +67,100 @@ def __init__(self, env=None, args=None):
+        if len(arg) == 2:
+          short_flags.add(arg)
+        elif kwargs.get('type') == bool:
+          flags.add('--no-{}'.format(arg[2:]))
+
+    GlobalOptionsRegistrar.register_bootstrap_options(capture_the_flags)
+
+    def is_bootstrap_option(arg):
+      components = arg.split('=', 1)
+      if components[0] in flags:
+        return True
+      for flag in short_flags:
+        if arg.startswith(flag):
+          return True
+      return False
+
+    # Take just the bootstrap args, so we don't choke on other global-scope args on the cmd line.
+    # Stop before '--' since args after that are pass-through and may have duplicate names to our
+    # bootstrap options.
+    bargs = filter(is_bootstrap_option, itertools.takewhile(lambda arg: arg != '--', self._args))
+
+    config_file_paths = self.get_config_file_paths(env=self._env, args=self._args)
+    config_files_products = yield config_file_paths
+    pre_bootstrap_config = Config.load_file_contents(config_files_products)
+
+    def bootstrap_options_from_config(config):
+      bootstrap_options = Options.create(
+        env=self._env,
+        config=config,
+        known_scope_infos=[GlobalOptionsRegistrar.get_scope_info()],
+        args=bargs,
+        option_tracker=self._option_tracker
+      )
+
+      def register_global(*args, **kwargs):
+        ## Only use of Options.register?
+        bootstrap_options.register(GLOBAL_SCOPE, *args, **kwargs)
+
+      GlobalOptionsRegistrar.register_bootstrap_options(register_global)
+      return bootstrap_options
+
+    initial_bootstrap_options = bootstrap_options_from_config(pre_bootstrap_config)
+    bootstrap_option_values = initial_bootstrap_options.for_global_scope()
+
+    # Now re-read the config, post-bootstrapping. Note the order: First whatever we bootstrapped
+    # from (typically pants.ini), then config override, then rcfiles.
+    full_configpaths = pre_bootstrap_config.sources()
+    if bootstrap_option_values.pantsrc:
+      rcfiles = [os.path.expanduser(rcfile) for rcfile in bootstrap_option_values.pantsrc_files]
+      existing_rcfiles = filter(os.path.exists, rcfiles)
+      full_configpaths.extend(existing_rcfiles)
+
+    full_config_files_products = yield full_configpaths
+    self._post_bootstrap_config = Config.load_file_contents(
+      full_config_files_products,
+      seed_values=bootstrap_option_values
+    )
+
+    # Now recompute the bootstrap options with the full config. This allows us to pick up
+    # bootstrap values (such as backends) from a config override file, for example.
+    self._bootstrap_options = bootstrap_options_from_config(self._post_bootstrap_config)
+
+  def construct_and_set_bootstrap_options(self):
+    """Populates the internal bootstrap_options cache."""
+    def filecontent_for(path):
+      with open(path, 'rb') as fh:
+        return FileContent(path, fh.read())
+
+    # N.B. This adaptor is meant to simulate how we would co-operatively invoke options bootstrap
```

Comment:
Nice. Yea, seems reasonable, although as mentioned above this will need to move out for now in order to not memoize options values across runs.

Note that `@rule`s can also recurse, which is sometimes more natural.

Reply:
fixed